lib/shoryuken/processor.rb

```diff
@@ -2,8 +2,77 @@ module Shoryuken
   class Processor
     include Util
 
-    def initialize(manager)
-      @manager = manager
+    BATCH_LIMIT = 10
+
+    def initialize(fetcher, polling_strategy)
+      @fetcher = fetcher
+      @polling_strategy = polling_strategy
+      # what is the worst thing that could happen because it's not atomic?
+      # is it possible that it evaluates to true even if we never set it to true?
+      @done = false
+    end
+
+    def start
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      # can it throw exception?
+      # what should we do if it dies?
+      @thread ||= spawn_thread(&method(:loop))
+    end
+
+    def terminate(wait=false)
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.value if wait
+    end
+
+    def kill
+      @done = true
+      # does updating and evaluating @thread has to be wrapped in mutex?
+      return if !@thread
+      @thread.raise ::Shoryuken::Shutdown
+    end
+
+    def alive?
+      # is this ok?
+      !!@thread.status
+    end
+
+    private
+
+    def loop
+      while !@done
+        do_work
+      end
+    rescue ::Shoryuken::Shutdown
+      # killed, bail!
+    rescue => e
+      # are we safe to retry here? What are kind of errors we do not want to recover from?
+      # OOM?
+      logger.debug { "Processor caught exception: #{e}" }
+      retry
+    end
+
+    def do_work
+      queue = @polling_strategy.next_queue
+      return logger.info { 'Doing nothing, all queues are paused' } if queue.nil?
+      batched_queue?(queue) ? dispatch_batch(queue) : dispatch_single_message(queue)
+    end
+
+    def dispatch_batch(queue)
+      batch = @fetcher.fetch(queue, BATCH_LIMIT)
+      @polling_strategy.messages_found(queue.name, batch.size)
+      process(queue.name, patch_batch!(batch))
+    end
+
+    def dispatch_single_message(queue)
+      messages = @fetcher.fetch(queue, 1)
+      @polling_strategy.messages_found(queue.name, messages.size)
+      process(queue.name, messages.first) if messages.length > 0
+    end
```

Comment:
@mariokostelac I think we can deprecated `delay` in favor of `wait_time_seconds`, if you don't want hold checking all queues, you can set it to zero, if you want to wait and save some $ (too little as you mentioned), you can bump it to whatever you would like to. 

> instead of the heartbeat interval we've discussed in #291

That implementation is straightforward and working. But as you raised, it isn't as optimal  as it could be, there will be always a possible 25ms delay, which isn't a huge thing for some cases, but I bet it's for others. If we need to wait, it feels like it's better to wait waiting for messages, not just for the heartbeat. But if we go with processors fetching messages, and long-pool (which feels cool indeed), I think we we will need the polling strategies to make it thread-safe, otherwise we won't be able to load balance the queues consumption, so we will move the complexity around, but we may ended with a very optimal setup.

I think Shoryuken doesn't need to address all needs. We can go with a solution that's good and straightforward as possible for most users and for us to maintain. Apps that can't wait for something less than second to process are very edge case.